### ISSUE TITLE
P1: Reduce RL policy-gradient normalization after near-zero trusted-gradient signal

### ISSUE THREAD

**Issue by lanyusea** [lanyusea/screeps#1319]
## Problem
Latest RL policy online advantage ledger reports the true-gradient computation path is now active, but the signal is flattened by the current gradient normalization scale:

- Policy advantage report: `runtime-artifacts/rl-control-loop/20260522T032832Z-policy-advantage.json`
- Ledger finding: candidate mode is `construction-priority.pg.{territory,resource,risk-aware}-seed.v1` / shadow; online utility remains `MIXED` for the 65th consecutive report.
- Regressions called out by the ledger:
  - `P1 — Gradient near-zero due to 1e9 normalization — FIX: Reduce normalization to 1e3-1e4`
  - `P1 — trustedGradientUpdate still false — gate correctly blocks, normalization fix should resolve`
- Training ledger: `runtime-artifacts/rl-control-loop/20260522T035208Z-training-ledger.json` reports `trainingDidRun=true`, 31 candidate evaluation iterations, 25 policy update iterations, and active gradient trust building.
- Current main after #1317: `b074628b`; an active Tencent validation run `tencent-pg-20260522t034423z` is still in progress, so additional paid compute should stay guarded, but code correction can proceed offline.

## Scope
Implement the smallest code/test change needed to make the RL policy-gradient update preserve informative gradient magnitude:

1. Replace or parameterize the hard-coded `1e9` gradient normalization factor with a scale in the `1e3`–`1e4` range, using repository naming/config conventions.
2. Add/update tests proving small but meaningful policy-gradient deltas are not flattened to near-zero while safety/trust gates still block genuinely untrusted updates.
3. Preserve existing scorecard, reward, and safety semantics; no official MMO writes and no paid compute launch from this issue.

## Acceptance
- Controller verification passes for the touched scripts/tests.
- A post-merge Loop A/B or guarded Tencent run shows non-near-zero gradient signal and `trustedGradientUpdate` can turn true when the trust criteria are met.
- Update #879/#1032/#1233/#1299 evidence after PR creation and after validation.

Refs #879 #1032 #1233 #1299

**Comment by lanyusea** [comment]
Scheduler dispatched the implementation lane for this issue.

- Worktree: `/root/screeps-worktrees/issue-1319-rl-gradient-normalization`
- Branch: `fix/issue-1319-rl-gradient-normalization`
- Codex session: `proc_c9c7af7df18f`
- OS pid: `1927731`
- Base: `origin/main` / `b074628b`
- Scope guard: code/tests only; no paid Tencent compute, no official MMO writes, no GitHub/Project mutation from Codex.
- Required verification in the Codex prompt: `py_compile`, `python3 -m unittest scripts.test_screeps_rl_training_runner -v`, and `git diff --check`.

Next scheduler action: inspect the worktree after Codex exits, verify the commit from the controller, push/open PR, add PR to Project, and then require post-merge Loop A/B validation.

**Comment by lanyusea** [comment]
#1320 review-fix checkpoint:

- Commit pushed: `4addda1c` (`fix(rl): preserve capped gradient update scale`), authored by `lanyusea's bot <lanyusea@gmail.com>`.
- Addressed active review findings via Codex triage (`/tmp/pr1320-codex-triage.md`): capped-normalization update-scale compensation and gradient-evidence export fields.
- Controller verification passed: `py_compile`, focused unittest (`95 tests OK`), and `git diff --check`.
- Current blocker: exact-head PR #1320 `prod-ci`/CodeRabbit/review-thread gate on `4addda1c`; do not dispatch another conflicting Codex lane in this worktree.
- Related compute state: Tencent run `tencent-pg-20260522t044005z` is active (pid 1960249, safety flags `liveEffect=false`, `officialMmoWrites=false`); no duplicate paid compute.

**Comment by lanyusea** [comment]
Issue #1319 implementation checkpoint (2026-05-22T05:37Z)

PR #1320 advanced to exact head `17c6274183b2414afcfbf2c946edaa5a0926ff2a` after two Codex review-fix commits:
- `d346af92` keeps `scalarRewardScaleFactor` full precision so extreme source weights cannot round the scale to `0.0` and suppress updates.
- `17c62741` applies the `10_000` normalization cap to the actual scalar update gradient instead of only recording it as evidence, matching #1319's goal to avoid 1e9 flattening.

Verified from the controller before commit:
- `git diff --check`
- `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m unittest scripts.test_screeps_rl_training_runner -q` (97 tests OK)

Current blocker: PR #1320 is still `In review` because CodeRabbit remains `PENDING` on exact head `17c62741` and review threads must be re-queried/resolved after it finishes. `prod-ci` is green.

**Comment by lanyusea** [comment]
Issue #1319 latest checkpoint (2026-05-22T05:50Z)

PR #1320 advanced to exact head `2c1bd77a6bd5e55cbecb593586cd07e579a47c7c`.

Latest review-fix details:
- Kept #1319's normalization-cap invariant: cap-normalized gradients drive actual offline update math; reverting to pre-cap source-max scaling was triaged as non-critical/owner-decision territory because it would reintroduce the 1e9 flattening this issue targets.
- Fixed a critical precision issue where rounded evidence could feed the actual update path. `2c1bd77a` now keeps unrounded scalar gradients/EMA values for policy update math while preserving rounded audit/display fields.
- Controller verification: `git diff --check`, py_compile for runner/tests, and `python3 -m unittest scripts.test_screeps_rl_training_runner -q` (98 tests OK).

Current blocker: PR #1320 is still in review because exact-head prod-ci and CodeRabbit restarted on `2c1bd77a`. Next pass should re-query checks/threads and merge only after the automated review gate clears.

**Comment by lanyusea** [comment]
Scheduler checkpoint dispatched Codex review-fix/triage for PR #1320.

- Worktree: `/root/screeps-worktrees/issue-1319-rl-gradient-normalization`
- Background session: `proc_6f3b8115f855` (OS pid `2109217`)
- Starting head: `2c1bd77a6bd5e55cbecb593586cd07e579a47c7c`
- Review-thread evidence: 6 unresolved total / 5 active non-outdated from `/tmp/screeps-scheduler-audit/pr-review-threads.json`
- Codex prompt requires per-thread classifications, `/tmp/pr1320-codex-triage.md`, `git diff --check`, `py_compile`, and `python3 -m unittest scripts.test_screeps_rl_training_runner -q`, with Codex-authored commit if a critical fix is required.

No push/merge will be attempted until the process exits, controller verification passes, exact-head checks rerun, and review threads are re-queried.

**Comment by lanyusea** [comment]
PR #1320 review-fix update (2026-05-22T06:21Z)

- New Codex-authored commit pushed: `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d` — `fix(rl): preserve raw EMA gradient state`
- Commit files: `scripts/screeps_rl_training_runner.py`, `scripts/test_screeps_rl_training_runner.py`
- Controller verification before commit: `git diff --check`; `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`; `python3 -m unittest scripts.test_screeps_rl_training_runner -q` → 99 tests OK
- Codex triage file: `/tmp/pr1320-codex-triage.md`

Triage summary:
- `PRRT_kwDOSMSsO86EA2h3`: FIX — raw EMA state was not round-trippable; patched by exporting rounded `emaGradient`, preserving full-precision `rawEmaGradient`, and consuming raw EMA for update math.
- `PRRT_kwDOSMSsO86EAhL8`: ADVISORY_ONLY — below-cap edge-case coverage is useful but not critical; no patch.
- `PRRT_kwDOSMSsO86EAmMI`: OWNER_DECISION / non-critical conflict with #1319 invariant; no revert to pre-cap scaling.
- `PRRT_kwDOSMSsO86EA2h5`: ADVISORY_ONLY — assertion/style coverage request, no critical defect after raw EMA fix.
- `PRRT_kwDOSMSsO86EA2h9`: ADVISORY_ONLY — metadata assertion request, no critical defect.
- `PRRT_kwDOSMSsO86EAXCQ`: RESOLVE_STALE_OR_OUTDATED — original scalarRewardScaleFactor zeroing issue is already fixed on current head.

Current exact-head gate after push:
- PR head: `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d`
- `prod-ci` is IN_PROGRESS.
- `CodeRabbit` is PENDING.
- GraphQL review threads after push: 5 unresolved total / 4 active non-outdated; do not resolve until CodeRabbit finishes and threads are re-queried on the exact head.

Next action: re-query exact-head checks and review threads after CodeRabbit/prod-ci complete; resolve fixed/advisory/stale threads with the triage evidence if no new critical findings appear, then run the merge gate.

**Comment by lanyusea** [comment]
PR #1320 latest-thread triage complete (2026-05-22T06:28Z)

- Codex latest-thread triage process: `proc_b17125e211c4` exited cleanly with no repo changes needed.
- Triage file: `/tmp/pr1320-codex-triage-latest.md`
- Worktree status: clean at head `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d`.
- Exact-head checks now pass: `prod-ci` SUCCESS and `CodeRabbit` SUCCESS.

Latest triage result:
- `PRRT_kwDOSMSsO86EAXCQ`: `RESOLVE_STALE_OR_OUTDATED` — scalarRewardScaleFactor zeroing allegation is fixed on current head.
- `PRRT_kwDOSMSsO86EAhL8`: `ADVISORY_ONLY` — below-cap edge-case coverage request is non-critical.
- `PRRT_kwDOSMSsO86EAmMI`: `OWNER_DECISION`/non-critical duplicate of the scale-invariance policy conflict; no code change because #1319 intentionally uses cap-normalized update magnitude as an offline/shadow gated fix.
- `PRRT_kwDOSMSsO86EA2h9`: `ADVISORY_ONLY` — metadata assertion request is non-critical.
- `PRRT_kwDOSMSsO86EBK8M`: `OWNER_DECISION`/duplicate of `EAmMI`; no code change.

Controller is resolving the triaged review threads and will run the merge gate only after a fresh GraphQL/PR/check/Project re-query.
